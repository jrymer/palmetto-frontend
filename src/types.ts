export interface Weather {
  description: string;
  humidity: string;
  main: string;
  pressure: string;
  temperature: Temperature;
  wind: Wind;
  visibility: string;
}

export interface Temperature {
  feelsLike: string;
  tempActual: string;
  tempMin: string;
  tempMax: string;
}

export interface Wind {
  direction: string;
  speed: string;
}

export interface Time {
  sunrise: number;
  sunset: number;
}

export interface Coordinates {
  lon: number;
  lat: number;
}

export interface WeatherPayload {
  coords: Coordinates;
  time: Time;
  weather: Weather;
  name: string;
}

export type GooglePrediction = google.maps.places.AutocompleteResponse;

export interface PredictionOption {
  display: string;
  value: string;
  id: string;
}

export interface WeatherQueryParams {
  search: {
    city: string;
    coords?: Coordinates;
  };
  unit: Units;
}

export type Units = 'imperial' | 'metric';

export const DEFAULT_CITY_SEARCH = 'Boulder, CO, USA';
