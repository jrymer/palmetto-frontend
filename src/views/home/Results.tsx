import React from 'react';
import styled from 'styled-components';

import { WeatherPayload } from '../../types';

const Container = styled.div`
  grid-area: Results;
  display: flex;
  flex-direction: column;
  height: -webkit-fill-available;
  min-width: 40%;
  width: var(--col-width);
  justify-content: space-between;
  margin: ${(props) => props.theme.margin};
  padding: ${(props) => props.theme.padding};
  border: solid thin ${(props) => props.theme.colors.offsetGrey};
`;

const MainTemperature = styled.div`
  font-size: var(--fs-600);
  font-weight: 600;
`;

const SubTemperature = styled.div`
  font-size: var(--fs-500);
`;

const WrapRow = styled.div`
  display: flex;
  flex-wrap: wrap;
  justify-content: space-between;
`;

interface ResultsProps {
  city: string;
  weather: WeatherPayload | undefined;
}

/**
 * Format a date value from epoch to UTC
 *
 * @param {number} epoch Epoch time
 * @return {string}  Time in HH:MM
 */
const getReadableTime = (epoch: number) => {
  const date = new Date(epoch);
  return `${date.getUTCHours()}:${date.getUTCMinutes()}`;
};

/**
 * Returns a result card with weather information about a city
 *
 * @param {string} city Selected city
 * @param {WeatherPayload | undefined} weather Weather payload from weather API
 * @return A result card component
 */
const Results: React.FC<ResultsProps> = ({ city, weather }) => {
  if (!weather) {
    return null;
  }

  const { weather: w, time } = weather;
  const { temperature: temp, wind } = w;
  const desc = w.description.charAt(0).toUpperCase() + w.description.slice(1);
  return (
    <Container>
      <div>{city}</div>
      <WrapRow>
        <div>
          <MainTemperature>
            {temp.tempActual} and {w.main}
          </MainTemperature>
          <SubTemperature>Feels Like: {temp.feelsLike}</SubTemperature>
          <SubTemperature>
            High: {temp.tempMax} Low: {temp.tempMin}
          </SubTemperature>
        </div>
        <div>
          <div>Humidity: {w.humidity}</div>
          <div>Pressure: {w.pressure}</div>
          <div>Visibility: {w.visibility}</div>
          <div>
            Wind: {wind.speed} at {wind.direction}
          </div>
        </div>
      </WrapRow>
      <WrapRow>
        <div>{desc}.</div>
        <div>
          Sunrise: {getReadableTime(time.sunrise)} Sunset: {getReadableTime(time.sunset)}
        </div>
      </WrapRow>
    </Container>
  );
};

export default Results;
