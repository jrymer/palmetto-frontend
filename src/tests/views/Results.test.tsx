import React from 'react';

import { WeatherPayload } from '../../types';
import Results from '../../views/home/Results';
import { createSnapshot, render, screen } from '../testWrapper';

const weater: WeatherPayload = {
  coords: {
    lat: 47.6062,
    lon: -122.3321,
  },
  time: {
    sunrise: 1651409512,
    sunset: 1651461645,
  },
  weather: {
    description: 'broken clouds',
    humidity: '57%',
    main: 'Clouds',
    pressure: '1014 hPa',
    temperature: {
      feelsLike: '15.61° C',
      tempActual: '16.42° C',
      tempMin: '14.89° C',
      tempMax: '18.29° C',
    },
    wind: {
      direction: '280°',
      speed: '3.09 mps',
    },
    visibility: '1000 km',
  },
  name: 'Seattle',
};

describe('Results tests', () => {
  it('Results snapshot is valid with base props', () => {
    const snapshot = createSnapshot(<Results city="Seattle" weather={weater} />);
    expect(snapshot).toMatchSnapshot();
  });
  it('Displays weather', () => {
    render(<Results city="Seattle" weather={weater} />);
    expect(screen.getByText(/seattle/i)).toBeInTheDocument();
    expect(screen.getByText(/16\.42° c and clouds/i)).toBeInTheDocument();
    expect(screen.getByText(/pressure: 1014 hpa/i)).toBeInTheDocument();
    expect(screen.getByText(/sunrise: 2:43 sunset: 2:44/i)).toBeInTheDocument();
  });
});
