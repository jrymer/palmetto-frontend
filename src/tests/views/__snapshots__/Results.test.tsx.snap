// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Results tests Results snapshot is valid with base props 1`] = `
<div
  className="sc-bczRLJ wnsic"
>
  <div>
    Seattle
  </div>
  <div
    className="sc-hKMtZM fvCfgH"
  >
    <div>
      <div
        className="sc-gsnTZi hTdNUT"
      >
        16.42° C
         and 
        Clouds
      </div>
      <div
        className="sc-dkzDqf iUQuVc"
      >
        Feels Like: 
        15.61° C
      </div>
      <div
        className="sc-dkzDqf iUQuVc"
      >
        High: 
        18.29° C
         Low: 
        14.89° C
      </div>
    </div>
    <div>
      <div>
        Humidity: 
        57%
      </div>
      <div>
        Pressure: 
        1014 hPa
      </div>
      <div>
        Visibility: 
        1000 km
      </div>
      <div>
        Wind: 
        3.09 mps
         at 
        280°
      </div>
    </div>
  </div>
  <div
    className="sc-hKMtZM fvCfgH"
  >
    <div>
      Broken clouds
      .
    </div>
    <div>
      Sunrise: 
      2:43
       Sunset: 
      2:44
    </div>
  </div>
</div>
`;
